Inject ReactiveContext as constructor argument

katis opened the pull request:
Made the ReactiveContext an explicit argument for classes that need it.

Builder functions provide the global context if it's not provided for them as an optional named argument.

I tried to implement the Zone local value for the context, but it did not work properly with tests. I don't know why exactly. It can be implemented in the future by changing `mobxContext` global to a getter that finds the current Zones context.

pavanpodila commented on lib/src/api/context.dart:
```diff
@@ -0,0 +1,4 @@
+import 'package:mobx/src/core/context.dart';
+
+/// Global Mobx context.
+final ReactiveContext mobxContext = ReactiveContext();
```
How about calling it `globalContext` or `defaultContext`?

katis replied:
I thought about this a little more, and if we want to explore the Zone thing at some point, maybe "currentContext" or "activeContext" would be a decent name? That way we can replace it with a getter if needed.

pavanpodila replied:
Yup that sounds good

pavanpodila replied:
let's go with `currentContext`

pavanpodila commented on lib/src/api/observable.dart:
```diff
@@ -47,15 +49,18 @@ ObservableValue<T> observable<T>(T initialValue, {String name}) =>
-Atom createAtom({String name, Function onObserved, Function onUnobserved}) =>
-    Atom(name ?? 'Atom@${ctx.nextId}',
-        onObserve: onObserved, onUnobserve: onUnobserved);
+Atom createAtom({String name, Function onObserved, Function onUnobserved}) {
```
No context argument here? Any reason why we are picking `mobxContext` directly?

pavanpodila replied:
Also the context should be a named argument

katis replied:
Yeah, missed this one.

pavanpodila commented on lib/src/api/reaction.dart:
```diff
@@ -56,8 +62,9 @@ ReactionDisposer when(
+  ReactiveContext context,
 }) =>
-    createWhenReaction(predicate, effect, name: name);
+    createWhenReaction(context ?? mobxContext, predicate, effect, name: name);
```
I think the context should go into the named argument list

pavanpodila replied:
I think that will keep the signature consistent throughout

katis replied:
I figured that `createWhenReaction` is an internal helper function, and as such the context should be a named argument. I've had trouble before in forgetting to pass this kind of stuff around if it's optional everywhere, so I prefer that just the public API's have it as an optional named argument.

pavanpodila replied:
ok that makes sense

pavanpodila commented on lib/src/core/action.dart:
```diff
@@ -2,14 +2,14 @@ import 'package:mobx/src/core/context.dart';
-  Action(Function fn, {String name}) {
-    _fn = fn;
+  Action(this._context, this._fn, {String name})
+      : name = name ?? _context.name('Action');
```
This is a good addition: using the `context.name()`👍

pavanpodila commented on lib/src/core/action.dart:
```diff
@@ -30,40 +30,15 @@ class Action {
       ..untrackedEnd(prevDerivation);
   }
 }
-
```
Why are we removing `runInAction()`?

katis replied:
It is now in api/action.dart

pavanpodila commented on test/observable_test.dart:
```diff
@@ -23,15 +24,15 @@ void main() {
 
   test('Raw observables', () {
-    final x = ObservableValue(1000);
+    final x = ObservableValue(mobxContext, 1000);
```
We will also need some tests that will provide custom context and verify its being used correctly.

pavanpodila replied:
I think that will apply to actions, observables, reactions

pavanpodila commented:
@katis Also from the next PR, you can use this repo with a separate branch. I made you a collaborator. That way I could also chip in if needed.

katis commented:
@pavanpodila made the changes requested, added tests using Mockito for context.
I added support for dev sdks in the version range too, since Flutter uses a dev version and my installation uses Flutters SDK.
